Docs: Refactor API Documentation Generator

jperson2000 opened the pull request:
## Description

This update refactors the API Documentation generator to add several features:

* Strongly typed documentation objects for `ApiPage.razor` to use
* Improved trimmability by removing all Reflection code from Blazor pages
* Support for "documentation coverage" metrics shown during builds

The new documentation generator has these features:

* Support for `<summary>` and `<remarks>` elements 
* Support for the `<c>` element (converted to `<code>`)
* Support for the `<para>` element (converted to `<p>`)
* Support for HTML in XML documentation comments
* Support for links to members in the same class or any base class
* Support for `<see cref` links

This update also adds Documentation Coverage metrics during builds:

```
XML Documentation Coverage for MudBlazor:

Types:      69 of 169 (41%) other types
Properties: 1458 of 2451 (59%) properties
Methods:    170 of 2093 (8%) methods
Fields:     12 of 667 (2%) fields/enums
Events:     0 of 6 (0%) events

API Builder: WARNING: 280 types have XML documentation which couldn't be matched to a type.
API Builder: WARNING: 569 properties have XML documentation which couldn't be matched to a type's property.
API Builder: WARNING: 653 methods have XML documentation which couldn't be matched to a type's property.
API Builder: WARNING: 4 events have XML documentation which couldn't be matched to a type's property.
API Builder: WARNING: 176 fields have XML documentation which couldn't be matched to a type's property.
```

This update also makes a few minor code cleanup changes:

* The `Paths` class is now static

![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/7f64b143-26c4-4810-9619-0216bf35f9ac)
![image](https://github.com/MudBlazor/MudBlazor/assets/18043079/7aef9ff0-414c-4006-b39a-ef3f7ab01b39)

## How Has This Been Tested?

The `TestsForApiPages` test generator was updated to use the new `Api` Blazor page when generating tests.  A big change is that we now test _all public types_, not just types inheriting from `MudComponentBase`.  This resulted in an additional ~400 tests being generated during builds.

Also, per #3595, an additional bank of tests are now generated which test the legacy API URL's.  This will ensure that users who have bookmarked the old API page will still work properly.  

The total API docs tests increased from ~150 to ~600.

## Type of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation (fix or improvement to the website or code docs)

## Checklist

- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

jperson2000 committed:
Docs: MudColorPicker: Added XML Documentation for public members

jperson2000 committed:
Revert "Docs: MudColorPicker: Added XML Documentation for public members"

This reverts commit 2ec8c975d1de05692f0cd83bc67b297fc6c88a7a.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into dev

jperson2000 committed:
Merge branch 'MudBlazor:dev' into dev

Commit:
Refactored API docs to support XML-to-HTML element conversion.  Support for some "see cref" links.

danielchalmers commented:
Just checking: does it support `<br/>`? I've been using that recently instead of `<para>` because it's more compact.

danielchalmers commented:
Just checking: does it support `<br/>`? I've been using that recently instead of `<para>` because it's more compact.

ScarletKuro commented:
Would you able to do this testing https://github.com/MudBlazor/MudBlazor/discussions/3595
To see that nothing unexpected got broken?

ScarletKuro commented:
> Would you able to do this testing #3595 To see that nothing unexpected got broken?

Maybe I'd be good to leave the code of this in the `tool` folder so anyone could run it at anytime.

iwis commented:
Why did you change `LayoutProperties.DefaultBorderRadius` to `DefaultBorderRadius` in the description of `MudChipset.Label`?

Some incorrect descriptions:
  `MudChipSet<T>.Comparer`
  `MudHighlighter.Markup`
  `MudMessageBox.YesButton`

iwis commented:
Why did you change `LayoutProperties.DefaultBorderRadius` to `DefaultBorderRadius` in the description of `MudChipset.Label`?

Some incorrect descriptions:
  `MudChipSet<T>.Comparer`
  `MudHighlighter.Markup`
  `MudMessageBox.YesButton`

I only checked a few pages manually. When I was refactoring the code for documentation API often some change was manifested only in a bug in a few properties on a few pages. This is why I wrote https://github.com/MudBlazor/MudBlazor/discussions/3595 to be sure I don't break documentation. As you see, unfortunately manual testing isn't enough :(

jperson2000 commented:
> Just checking: does it support `<br/>`? I've been using that recently instead of `<para>` because it's more compact. para might want a little more space in the docs compared to br (but much less than before), what do you think?

I'm pretty sure `<br/>` and other tags would be supported since the new code uses `MarkupString`.  I'll be sure to test that.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 commented:
> Why did you change `LayoutProperties.DefaultBorderRadius` to `DefaultBorderRadius` in the description of `MudChipset.Label`?
> 
> Some incorrect descriptions: `MudChipSet<T>.Comparer` `MudHighlighter.Markup` `MudMessageBox.YesButton`
> 
> I only checked a few pages manually. When I was refactoring the code for API documentation, often some change was manifested only in a bug in a few properties on a few pages. This is why I wrote #3595 to be sure I don't break documentation. As you see, unfortunately, manual testing isn't enough :(

Thanks for checking some pages!  I agree, it's really tough to get every detail just right, like dealing with reflection names.  Hopefully I can get some automated tests built from your great test code.

Commit:
Progress on a new XML documentation compiler

Commit:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

Commit:
Reverted DocStrings so we can roll back if needed

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

Commit:
dotnet format.  Re-enabled DocStrings for now

ScarletKuro commented on src/MudBlazor.Docs.Compiler/Program.cs:
```diff
@@ -10,7 +11,7 @@ public static int Main()
             var success =
                 new CodeSnippets().Execute()
-                && new DocStrings().Execute()
+                && new MudComponentDocumenter(typeof(MudText).Assembly).Execute()
```
I would probably use the `_Imports` file to get the assembly, at least it's what MS practicing.

jperson2000 replied:
Good catch, I've changed to `_Imports`  (in the class)

ScarletKuro commented on src/MudBlazor.Docs.Compiler/DocumentedEvent.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public string Summary { get; set; }
+    public string Remarks { get; set; }
+    public bool IsEmpty => string.IsNullOrEmpty(Summary) && string.IsNullOrEmpty(Remarks);
```
I'd recommend to add `MemberNotNullWhen` attribute. For others too

jperson2000 replied:
I ended up pulling out the `IsEmpty` properties, but let me know if you see good places for this.  This attribute is new to me

Commit:
Progress on newer XML documentation generator.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

Commit:
Progress on API Docs page with MudChips for categories and keyword search.

Commit:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

Commit:
Minor fixes before ending for the day

Commit:
Progress before switching to another PR

Commit:
Added type hierarchy component.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

danielchalmers commented:
Wowww this is really impressive! Using native MudBlazor controls for it is a great idea.

Even having the icons for MudBlazor classes is so cool
![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/82b68be7-c885-44a5-b102-533005a9b0fe)

---

Few rough ideas from looking around (I know it's still a WIP):

- API tree doesn't need to be collapsible so I would keep it structurally the same but make the only clickable part the titles themselves. I think this is more intuitive.

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/3208058a-8c2d-4718-81fb-cf9fff4c0d99)

 

- Chips need fixed width so they dont shift. I dont even know if we have an api for that though

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/7dcaedaf-94ae-40b1-bcea-101105f0a0c3)
![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/d750e520-766f-4551-88f1-9d104b80fb4b)

 

- Should it use the language types instead of runtime so there's no mismatch from what we actually use?

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/71965764-5480-4518-a26f-6dee1f090eda)

 

- Needs to be denser to fit more on screen

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/44e4854a-6e09-473d-9d73-888de2449fb0)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/ebf4f684-09c9-49ec-af1e-67b4385dd7f5)

 

- tooltips are not aligned because of #1167 i think

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/ed6f7494-6acd-41d2-8fc4-12ef0e824e65)

 

- is there any way to get to the regular page like the old link?

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/a154f1f7-4d76-4df0-8f31-356e2a994f19)

  

- Should links be same typo as regular code blocks? (currently links are bigger)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/5ea4ec87-8f90-4d5b-a217-dffe84dfcefd)

 

- Can the links be constructed as href so we can hover and see where it goes + middle click to open in new tab?
- Should the inheritence filter be disabled by default so we can see more? the info it gives doesn't feel valuable enough to be there by default
- \>\>\> Need to make sure everything's accessible for screen readers, etc

danielchalmers commented:
One small concern I have about the code is that I see a lot of hardcoded strings; Can it find the info automatically?

jperson2000 commented:
> Few rough ideas from looking around (I know it's still a WIP):

Oh wow, thanks for taking the time to put all of this together!   All of it great feedback

jperson2000 commented:
> One small concern I have about the code is that I see a lot of hardcoded strings; Can it find the info automatically?

Are you referring to the dictionary that maps API URL's to types?   If so, then yes, there's a challenge with links because I want to be backwards compatible with anyone who has bookmarked an API page, but it means that a mapping dictionary is required to convert "alert" into "MudAlert" and etc.   The new docs will accept both "api/alert" and "api/MudAlert" URL's though if we wanted to gradually move to actual types in URL's.

There's a "critical types" list which is just a whitelist of the primary MudBlazor components, but this is only used to calculate documentation coverage metrics.   We could get rid of it and calculate docs coverage for ALL types, but this would include types users don't see and be a pretty low percentage.  Totally optional, though.  Thanks again for the detailed review!

Commit:
Switching branches

danielchalmers commented:
> Are you referring to the dictionary that maps API URL's to types?

That makes sense, thanks. I wonder if we could have some central location for all this. If you check `MenuService.cs` you'll see my concern.

danielchalmers commented:
The search and and filters are super cool (and I don't wanna be a killjoy) but I'm wondering if it's better to keep it simple - Most components don't have too many properties and I'm unsure if people would use it over Ctrl+F etc. And more moving parts = more places to go wrong. If it were like 100s of properties it'd make sense

henon commented:
> There's a "critical types" list which is just a whitelist of the primary MudBlazor components, but this is only used to calculate documentation coverage metrics. We could get rid of it and calculate docs coverage for ALL types, but this would include types users don't see and be a pretty low percentage.

Wouldn't a blacklist be better suited here? So when we add a component we don't have to think about extending the white list. Of course, if we add internal support types we have to remember, but even if we forget, the worst that can happen is that an internal type is exposed in the api docs for a time until we black list it.

henon commented:
I love that you added the search feature, but yeah, thinking again browser search pretty much does the same. When I said I'd love integration of the API info with search I meant the primary search box in the app bar. So for instance, if you type a parameter name that components will be listed that have that parameter. Not sure if you understood it this way.

Commit:
Fixed up page from PR comments

Commit:
Removed "Critical Types" concept

Commit:
Changed to use "ApiLink" for link calculation.  Removed hardcoded strings.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

Commit:
Fixed API linking from examples

Commit:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

danielchalmers commented:
FYI getting this error from `/api/mudalert` or `/api/alert` (or any other) on the `MudBlazor.Docs.WasmHost` project

```
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Object reference not set to an instance of an object.
System.NullReferenceException: Object reference not set to an instance of an object.
   at MudBlazor.Docs.Models.ApiDocumentation.GetType(String name) in \src\MudBlazor.Docs\Models\Generated\ApiDocumentation.cs:line 29
   at MudBlazor.Docs.Pages.Api.Api.OnParametersSet() in \src\MudBlazor.Docs\Pages\Api\Api.razor.cs:line 33
   at Microsoft.AspNetCore.Components.ComponentBase.CallOnParametersSetAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
```

---

Some more ideas since I think my review was requested. Again, **just my thoughts**, **not demands** 😅 

### the breadcrumbs could replace Inheritance as it's much more compact

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/e02597a5-954d-4b3e-b61c-c817dc730030)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/b7dda80f-cd43-40cb-8d16-88f51bfb55cd)

(remove ^)

### Breadcrumbs position

could go above the header

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/1c8b6a5e-d41c-4924-a032-249850cf9276)

### Breadcrumbs accessibility

https://developer.mozilla.org/en-US/docs/Web/CSS/Layout_cookbook/Breadcrumb_Navigation#accessibility_concerns

should have some kind of indicator for hover? (im not sure)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/036f25f4-25f5-407f-ba2a-c7cda496dd81)

### Breadcrumbs active

the ones that aren't active should have a different color, maybe just darker grey.

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/7ea62166-fbb2-4d8c-9a5b-241d0a9a47dc)

Better ^

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/8ad29563-cd69-47b6-b4ba-9bcb72672d66)

Worse but an option (primary color) ^

### should the headers look like they're a part of the table? (they have lines)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/800a1bd0-1107-4ff0-8305-517b33591749)

also i wonder if each header should have their own sort so it's always accessible? that would kill a few birds with one stone

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/d1b32f22-6eb4-4215-894c-aadb0ac59f95)


![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/188a71c6-834c-4439-84d2-65c8ea9efad3)


### the space between each section could be reduced

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/b689e623-70f3-4c80-81c7-1268bf3c0d5f)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/e2f7af4b-3fed-4bbd-bab9-1459f55772f0)

### Links could be default instead of Body1 so they look like normal links

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/3842be27-c493-4752-af92-52ddf536893e)

### subtext

subtext could semantically be subtitle1 instead of body1 (slightly different)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/a1cb7e69-3680-472e-8280-743f8cfe1447)

### Filters

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/2af6a0a9-908a-4154-9b8b-7c01b5bccc58)

Personally would try to simplify as much as possible so removing these and moving the "Inherited" info into the rows directly might work in order to save more space

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/eadaa8c1-9207-47e4-b1aa-0d2a152f3556)

Or better: an icon next to it that indicates there is a parent and when you hover over it it tells you where it comes from



If you still wanted to be able to collapse them, a simple clickable collapse arrow could do the trick

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/87247069-c057-4a8c-8d02-2a8b71ebdd76)

basically with the goal of keep the screen clean with less elements so people focus directly on the docs as their first instinct. dont want to scare anyone away because of extra buttons etc (sounds stupid i know lol)

### Accessibility

This area in the docs should have aria-labels for relevant areas and properties should be tabbable (currently all are skipped over). im aiming for accessibility as a priority now

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

Commit:
Progress on new API docs.  Improved support for methods (more to do).  Support for derived types.  Fixed some PR review suggestions.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
MudGlobal: Added support for linking global settings to their related components.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Improved support for generics.  Optimized memory usage for in-memory API objects.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Fixed API unit test generator to use the new "Api" page.  (The same components are tested as before.)

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
dotnet format fix, and fixed project to use new "ApiDocumentation" generated file instead of "DocStrings"

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 committed:
Changed generated tests to test ALL public types.  Tests now ensure that documentation was actually found, and that a link to the example exists for components.  Added ~400 tests.

jperson2000 commented:
Alrighty, after almost a month, I think this Docs refactor is nearly ready to review!  There are some things I still have to improve like documenting method parameters and enums/fields, but I think the site is in good shape to look at.  Here are the major differences to look for:

* The API docs exist for _all public types_, not just types inheriting from `MudComponentBase`.  This helps to cover non-component documentation such as `TableState` and others.
* The API test generator now generates tests for all public types.  The tests assert that documentation was actually found for the page.  For types which are Mud components, the tests also assert that an example link was found.

Let me know if you'd like any UI changes, and I'll bring this out of draft once I have the last pieces done in about 2-3 days.  Thanks @ScarletKuro @henon @danielchalmers @mckaragoz @Anu6is for your valuable feedback as always!

#### Steps to Test

1. From this branch, run `MudBlazor.Docs.Compiler` to generate new docs and tests on your machine.
2. Change the startup project to Server or WASMHost and run.
3. Navigate to any API component page.

ScarletKuro commented on src/MudBlazor.Docs/Models/Generated/DocumentedMethod.cs:
```diff
@@ -20,7 +20,27 @@ public class DocumentedMethod
+    {
+        get
+        {
+            if (string.IsNullOrEmpty(DeclaringTypeName))
```
Can we move this code in a reusable utility? This and code in the `DocumentedProperty.cs` look the same.

jperson2000 replied:
Great catch, @ScarletKuro, I ended up making all of the member types like `DocumentedProperty`, `DocumentedMethod` and etc. inherit from a new `DocumentedMember` which has all of the common properties and logic.

danielchalmers commented:
Thanks for taking into account so much of my feedback, I appreciate that a lot :)
Everything looking great, big improvement over the v6 docs.

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/99176e1d-4ae3-4815-afa7-ad31c9ab8179)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/c7c00d98-b149-4169-ab04-231bc380391c)

On mobile:

- Horizontal padding limits amount of info that can be seen
- Header is right-to-left (normally it's on the left side)
- The select ("Name") doesn't have a hint about its function
- May need very slightly more vertical padding between rows so they don't blend into each other

Accessibility (cc @igotinfected what do you think? screenshots above):
- Should you be able to tab between rows?
- Contrast between code blocks and background is low in dark mode
  - Swapping the row hover and default background color might work
  - Also might help to replace the code blocks with simple color text (aka remove the background)
- Aria labels ("Property name", "Property type", "Property description")?

Other:
- I think I was wrong when I said to skip documenting `RightToLeft` because I didn't think about it being in the docs

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Simplified documentation objects per @ScarletKuro.  Fixed PR issues reported by @danielchalmers.  Improved accessibility and display on mobile devices.

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 commented:
> Suggestions for improvements

Great catches as always, @danielchalmers !  I removed the `pa-8` on tables for better display on mobile devices, and added `aria-label` for table headers.  The right-aligned table group seems like the `MudTable` default for `xs` sizes.

jperson2000 committed:
dotnet format

jperson2000 committed:
Added additional tests for legacy API URLs based on discussion #3595.   Will fix code to pass tests soon.

jperson2000 commented:
@henon I added API docs tests for all of the legacy API URL's based on #3595.  Thanks again for the reminder, and I still have your suggestion on my list to add public types to the main MudBlazor search as part of this PR.

henon commented:
Great, I'll check out your code and run it for a thourough review soon.

iwis commented:
Below is not a comprehensive review, but only a report of a few issues that I noticed without further analysis.

Things that are less user-friendly in new doc:
 * In the previous UI, if I wanted to know what is the difference between https://mudblazor.com/api/textfield and https://mudblazor.com/api/numericfield I switched from "Categories" to "Inveritance" and I saw that:
   MudTextField       has additional fields: Clearable, AutoGrow, InputType, Mask, MaxLines;
   MudNumericField<T> has additional fields: Clearable, HideSpinButtons, InvertMouseWheel, Min, Max, Step.
 That way I was able to leanr about these two compoenent easily.
 In the new UI you have removed the "Categories/Inheritance/None" switch, so how can I learn this in the new UI?
 * On https://localhost:5001/api/MudAutocomplete%601 it would be better if the properties categories will be in the old order:
     Data
     Validation
     Behavior
     Appearance
     List behavior
     List appearance
     Common
   The order has the crucial meaning - more meaningful than the alphabetical one.
 * On https://localhost:5001/api/MudColorPicker there are is "Default" column in the "Properties" table. So for example one cannot simpy find out which icon is it. On https://mudblazor.com/api/colorpicker you can see the look of this icon, so it is easier to connect it to the real piece of the Color picker component that you have seen in the examples.
 * On https://localhost:5001/api/MudAutocomplete%601 there is no informtion about the elements of enumeration type "Margin" in the "Margin" proeprty (None, Dense, Property).
 * On the https://dev.mudblazor.com/api/autocomplete there is a clock "Rendered in 138 ms". I dont see such clock on https://localhost:5001/api/MudAutocomplete%601, so we cannot compare if the pages are fast.

Bugs:
 * On https://localhost:5001/api/Bar the link in "Examples" isn't correct.
 * On https://localhost:5001/api/MudColorPicker there are no descriptions for many properties e.g. "Adornment". Did you run the https://github.com/MudBlazor/MudBlazor/discussions/3595 ??
 
 * On https://localhost:5001/api/MudFormComponent%602#methods in the "MEthods" section the name of the "Return type" column should be not splited into two rows.
 * On https://localhost:5001/api/MudDebouncedInput%601 in the "Touched" property there is "`2" in the "MudFormComponent`2.RequiredError".

On https://localhost:5001/api/MudAutocomplete%601 why there are event callbacks "OnKeyDown" and "OnKeyUP" in "Properties" section and not in "EventCallbacks" section?? why did you remove the EventCallbacks section??

Generally you removed my two improvements:
  https://github.com/MudBlazor/MudBlazor/issues/2823
  https://github.com/MudBlazor/MudBlazor/pull/3172

iwis commented:
Below is not a comprehensive review, but only a report of a few issues that I noticed without a longer analysis, which is still needed.

Things that are less user-friendly in the new doc:
 * In the previous UI, if I wanted to know what is the difference between https://mudblazor.com/api/textfield and https://mudblazor.com/api/numericfield I switched from "Categories" to "Inheritance" and I saw that:
   MudTextField       has additional fields: Clearable, AutoGrow, InputType, Mask, MaxLines;
   MudNumericField<T> has additional fields: Clearable, HideSpinButtons, InvertMouseWheel, Min, Max, Step.
 That way I was able to learn about these two component easily.
 In the new UI you have removed the "Categories/Inheritance/None" switch, so how can I learn this in the new UI?
 * On https://localhost:5001/api/MudAutocomplete%601 it would be better if the properties categories are in the old order:
     Data
     Validation
     Behavior
     Appearance
     List behavior
     List appearance
     Common
   The order has a crucial meaning - more meaningful than the alphabetical one.
 * On https://localhost:5001/api/MudColorPicker there is no "Default" column in the "Properties" table. So for example one cannot simply find out which icon it is. On https://mudblazor.com/api/colorpicker you can see the look of the icon, so it is easier to connect it to the real piece of the Color picker component that you have seen in the examples.
 * On https://localhost:5001/api/MudAutocomplete%601 there is no information about the elements of enumeration type "Margin" in the "Margin" property (None, Dense, Property).
 * On the https://dev.mudblazor.com/api/autocomplete there is a clock "Rendered in 138 ms". I don't see such a clock on https://localhost:5001/api/MudAutocomplete%601, so we cannot compare if the pages are slower in the new UI. This is bad. In the beginning, the pages were loading in 2 seconds. We need the counter to monitor the loading times - we don't want them to load again in two seconds and the 2 sec time is very easy to obtain without the counter,

Bugs:
 * On https://localhost:5001/api/Bar the link in "Examples" isn't correct.
 * On https://localhost:5001/api/MudColorPicker there are no descriptions for many properties e.g. "Adornment". Did you run the https://github.com/MudBlazor/MudBlazor/discussions/3595 ??
 
 * On https://localhost:5001/api/MudFormComponent%602#methods in the "Methods" section the name of the "Return type" column shouldn't be split into two rows.
 * On https://localhost:5001/api/MudDebouncedInput%601 in the "Touched" property there is "`2" in the "MudFormComponent`2.RequiredError".

On https://localhost:5001/api/MudAutocomplete%601 why are there event callbacks "OnKeyDown" and "OnKeyUp" in the "Properties" section and not in the "EventCallbacks" section?? why did you remove the EventCallbacks section??

Generally, you removed my two improvements:
  https://github.com/MudBlazor/MudBlazor/issues/2823
  https://github.com/MudBlazor/MudBlazor/pull/3172
but why?

To sum up, in my opinion, the previous UI is more user-friendly and less buggy than the new UI.

henon commented:
> * On https://localhost:5001/api/MudAutocomplete%601 it would be better if the properties categories are in the old order:
>   Data
>   Validation
>   Behavior
>   Appearance
>   List behavior
>   List appearance
>   Common
>   The order has a crucial meaning - more meaningful than the alphabetical one.

I think this needs more argumentation why this order is supposed to be so meaningful. I supect it is based on personal preference. Others may say that Appearance, List appearance, Behavior, List behavior, Data, Validation, ... etc is a better order. In the end, alphabetical is a good order if people can not agree on a "correct" order.

iwis commented:
Sorry, I meant the https://github.com/MudBlazor/MudBlazor/discussions/9077 above.

jperson2000 commented:
> Below is not a comprehensive review, but only a report of a few issues that I noticed without a longer analysis, which is still needed.

Thanks for the valuable feedback, @iwis !  This is a draft PR with a ground-up rewrite of the docs compiler so that we can access documentation from anywhere in the MudBlazor docs web site via models, and without Reflection.  Thus, a lot of work has to be done to restore features like the ones you mentioned.  Please don't think I'm "taking features away" from you; on the contrary, I'm taking all feedback like yours into account to make sure this isn't a step backwards when it's finally ready for a formal review in a few weeks.  I'll tag you for additional feedback once I get more features implemented.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Excluded _Imports from docs

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 committed:
Improved support for properties, fields, and cross-referenced links.

mikes-gh commented:
@jperson2000 If you're changing the name of the generated code you may need to look at the project files IncludeGeneratedFiles targets and cleanup targets

jperson2000 committed:
Fixed new API docs tests for legacy api/ links

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Attempting to fix build

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 committed:
Refactored Type hierarchy tree to use new TreeItemData<T>

jperson2000 committed:
Improved ApiText component to render MudTooltip, MudLink, and MudText components.

jperson2000 committed:
dotnet format

jperson2000 committed:
Improved support for inline icons, API links, and external links.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 commented:
> @jperson2000 I wonder how much work this would be to change to a source generator

That's a good question!  I've never written one before but I bet the `ApiDocumentationWriter` could be turned into one  🤔  The [Source Generator docs](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) don't look too complicated.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/docs-generator-tagsupport

ScarletKuro commented:
> @jperson2000 I wonder how much work this would be to change to a source generator

I don't think a source generator is necessary for API documentation. The source generator works while you edit code in IDE(if I'm not wrong), which is unnecessary. We just need to create the file during compilation, and using the `Microsoft.CodeAnalysis.CSharp` API would be enough, just like it was done for the new icon generator: [GoogleMaterialDesignIconsGenerator](https://github.com/MudBlazor/MudBlazor.Icons/blob/master/src/GoogleMaterialDesignIconsGenerator/Generator/GeneratorBase.cs).

ScarletKuro commented:
> > * On https://localhost:5001/api/MudAutocomplete%601 it would be better if the properties categories are in the old order:
> >   Data
> >   Validation
> >   Behavior
> >   Appearance
> >   List behavior
> >   List appearance
> >   Common
> >   The order has a crucial meaning - more meaningful than the alphabetical one.
> 
> I think this needs more argumentation why this order is supposed to be so meaningful. I supect it is based on personal preference. Others may say that Appearance, List appearance, Behavior, List behavior, Data, Validation, ... etc is a better order. In the end, alphabetical is a good order if people can not agree on a "correct" order.

I kind of understand "Data" being the first one since it's basically the most important, and when you're prototyping something, you need it the most. The rest is pretty much noise as the appearance and validation can be done afterward.

I would say that filtering could solve all the "preference" issues. For example, having a checkbox for the categories, allowing you to hide or unhide the things you need. However, that might be overkill.

iwis commented:
@ScarletKuro, could you also consider the MudTable example here: https://github.com/MudBlazor/MudBlazor/discussions/9077?

ScarletKuro commented:
> @ScarletKuro, could you also consider the MudTable example here: #9077?

I would order it slightly differently, but my main point is that having a subjective order is better than having no order at all or just an alphabetical one.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
More progress on docs generator.  Appeasing SonorLint analyzers.   Added type names to @page directives to eliminate need for lookup tables.

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

danielchalmers commented:
Looking great, just small things (many could be delayed for a future release):

MudBooleanInput is shown as MudboolInput (and others like MudInputstring)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/77430c0c-e531-4410-b9a4-8004c1e99782)

Tooltips duplicate information (see left and right)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/fa7394bb-e56b-4052-8d6a-597ff54090c4)

Some tooltips may be too long

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/7136b4eb-ac8c-4e30-beac-917246d7956e)

String is linked and takes you to `api/String` (regular types shouldnt have coverage?)

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/0608d3e8-7460-4b56-abac-ec7d75ebc6d5)

It shows a link and a tooltip for base types

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/0ef548ba-6fcd-41ac-adf8-a03b75716ba1)

The base types have an example link that 404s

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/356e4d09-9095-45cc-891b-4a8c50f00ff1)

Base members shouldn't be listed

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/ed50ab52-b586-4e5c-a179-27dfe4e1d1f8)

Font size is too far apart with h3 and caption being large and very small

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/5a3c47d6-e85b-43ec-b19c-510a6eee77df)

Sort buttons aren't functional

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/81c98622-9c03-4310-922d-92324224bb68)

A single long row text makes it hard to see summaries especially on small screens. Maybe force a max width and wrap them?

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/906248f1-5733-4ba6-ac9a-ed0c811872a5)

The examples should probably stick out more so people recognize them better

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/bbddb84c-0d12-4460-a430-90b84a096d4d)

See MUI

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/7580fdad-7ba5-4094-98c8-dc1ee367b999)

The API Index is matched for the Component API nav menu so it should probably have a link of its own at the very top

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/0a91c689-96e7-4408-b468-1f9aa7f7671e)

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
Fixed issues reported by @danielchalmers.  Improved linking to and from "api/" and "component/" pages.

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 committed:
Merge branch 'dev' into feature/docs-generator-tagsupport

jperson2000 committed:
dotnet format --i-forget-this-every-time

jperson2000 committed:
Merge branch 'feature/docs-generator-tagsupport' of https://github.com/jperson2000/MudBlazor into feature/docs-generator-tagsupport

jperson2000 committed:
Fixed issue where API Member Table didn't look good in mobile-device mode

jperson2000 committed:
Rolled back unintended solution change.

danielchalmers reviewed (approved):
LGTM! Really impressive PR, glad we're getting this in for v7. I haven't done a full code review but I like the UI. There are some things I notice but we can discuss it for v7.1.0 or I'll try to PR myself and tag you for review instead of making you run around

danielchalmers commented:
@henon @ScarletKuro @mikes-gh @igotinfected  If you think this is ready we can release v7 this weekend. Remember we can update it afterwards for v7.1.0

igotinfected commented:
Not fully up to speed once again, noticed some minor issues but feel free to ignore:

- should the `.Docs` project automatically build `.Docs.Compiler` so we don't have to manually rebuild the solution or that specific project to launch the docs projects?
- when going to the examples page from the API page the side bar listing the contents and the link to the API do not appear (and the URL is different than the one from the left navigation menu):

![image](https://github.com/MudBlazor/MudBlazor/assets/15004223/6906323f-b76b-4514-99ec-2d6470e9ddb9)

via API URL:
![image](https://github.com/MudBlazor/MudBlazor/assets/15004223/6ba67a78-ced7-411f-9f83-c9f2c00be5e0)

via navigation URL:
![image](https://github.com/MudBlazor/MudBlazor/assets/15004223/31b0c295-3258-4e68-8552-579875d91b9a)

another example is the `MudAvatarGroup` examples page URL links to `MudAvatarGroup`, shows the `Avatar` page, and has no API URLs either

---

other than that LGTM!

ScarletKuro commented:
> @henon @ScarletKuro @mikes-gh @igotinfected If you think this is ready we can release v7 this weekend. Remember we can update it afterwards for v7.1.0

I see that not all of @iwis' concerns were addressed, such as:

1. Missing the Categories/Inheritance/None. Do we plan to add it back?
2. Missing the render time in debug; all our docs pages have this metric.
3. No `EventCallback` section anymore?
4. Did we decide on the ordering? Some said whatever but they are also people that preferred the old sorting (or similar) over the alphabetic one.

From me:
1. It shows internal properties like `ParameterContainer`, `InputIdState` when they shouldn't show up.
2. Maybe it's just me, but I miss the sign that indicates a parameter can be two-way bindable. It catches your attention and you don't need to look for `XyzChanged`.

jperson2000 commented:
> * when going to the examples page from the API page the side bar listing the contents and the link to the API do not appear (and the URL is different than the one from the left navigation menu):

Thanks, @igotinfected !  This appears to be a bug in the existing docs components where the component isn't detecting that the page has changed.  This is fixable with some code during `OnParametersSet` which I can fix in the next PR.

jperson2000 commented:
> I see that not all of @iwis' concerns were addressed, such as:

Yes, I haven't forgotten and I consider the feedback important, but I wanted to get this PR to a good state for the 7.0 release presumably tomorrow, then I'd open a new PR to keep going.

> 1. Missing the Categories/Inheritance/None. Do we plan to add it back?

Yes, it can be added back.

> 2. Missing the render time in debug, all our docs pages have this metric.

To me, the render time doesn't have a lot of value to end users, and may be moot now that no Reflection is being performed. 
 There will be an initial ~1s one-time load for all documentation objects, after which the render will be instant (10-20ms) for all pages.  If people want it back in, though, and think it has value to end users, I'll add it back.

> 3. No `EventCallback` section anymore?

This is something I do plan to improve.  `EventCallback` is a _property_ as far as Reflection and XML docs are concerned, so it takes a bit of filtering to make these properties show in the "Events" section along with actual `event` delegates. 

> 4. Did we decide on the ordering? Some said whatever but they are also people that preferred the old sorting (or similar) over the alphabetic one.

I did get the ordering finished, yes.  The `Order` and `Category` are now set from the `CategoryAttribute` and are part of the table sort.   The specific sort order is:  Order > Category > Table Column Sort when grouping by Category.

> 1. It shows internal properties like `ParameterContainer`, `InputIdState` when they shouldn't show up.

One big improvement to docs is that _all public types_ are now documented, which greatly increases the number of pages (and visibility for SEO), but it does expose the internal-use types which were marked as `public` instead of `internal`.   There is an `ExcludedTypes` variable which we can add types like this to to clean it up.

> 2. Maybe it's just me, but I miss the sign that indicates a parameter can be two-way bindable. It catches your attention and you don't need to look for `XyzChanged`.

Sure, this can be added in the next PR.  

> > should the `.Docs` project automatically build `.Docs.Compiler` so we don't have to manually rebuild the solution or that specific project to launch the docs projects?
> Wasn't that the case before? But yea I noticed that I need to compile the api docs first before running the website.

I remember having to run the compiler at least once locally to generate `DocStrings.generated.cs` before this PR, so this behavior is the same.  It was mentioned to perhaps use a Source Generator instead of the `ApiDocumentationWriter` class so that this is always built during the compile.  

Thanks as always for the reviews, and let me know what's most critical for the 7.0.0 release, otherwise I'll definitely address everything post-release.

ScarletKuro commented:
> To me, the render time doesn't have a lot of value to end users, and may be moot now that no Reflection is being performed.

It is not for end users, in fact, the end user doesn't see it. It only shows up in local or dev builds. I was thinking more of it as a way to check whether the performance has degraded, improved, or remained the same with each new feature. But I will let @henon have a word on that.

> but it does expose the internal-use types which were marked as public instead of internal

But `ParameterContainer` is `readonly internal`:
https://github.com/MudBlazor/MudBlazor/blob/2558c4eb865b12b2a8125e5090b358b30c26b5da/src/MudBlazor/Base/ComponentBaseWithState.cs#L18
No one is able to use it outside of the MudBlazor source code, so is there a point in showing it? My personal opinion is that it shouldn't be shown.

> then I'd open a new PR to keep going.

Ok, good to know!

jperson2000 committed:
Re-added <DocsRenderBenchmark /> to measure page performance

jperson2000 commented:
> It is not for end users, in fact, the end user doesn't see it. It only shows up in local or dev builds. I was thinking more of it as a way to check whether the performance has degraded, improved, or remained the same with each new feature. But I will let @henon have a word on that.

Ohh, I understand now.  Thank you.   Okay, that's been added back.

iwis commented:
The first bug: The description of the "For" property on https://localhost:5001/api/MudAutocomplete%601 is incorrect: "... When using an <MISSING>, this property is used to find data annotation validation attributes such as <MISSING> used to perform validation."

jperson2000 commented:
> The first bug: The description of the `For` property on https://localhost:5001/api/MudAutocomplete%601 is incorrect:

Thanks, @iwis.  Looks like an issue when a `see cref` goes to a .NET type, in this case `<see cref="EditForm"/>` (`Microsoft.AspNetCore.Components.Forms.EditForm`).

iwis commented:
Also, there are still bugs reported in the https://github.com/MudBlazor/MudBlazor/pull/8903#issuecomment-2099194694, and the bug:
> On https://localhost:5001/api/MudColorPicker there are no descriptions for many properties e.g. "Adornment".

reported in the https://github.com/MudBlazor/MudBlazor/pull/8903#issuecomment-2134032168.


If on the https://localhost:5001/api/MudDebouncedInput%601 page you click on the `RequiredError` link in the description of the `Touched` property then you have to scroll up a little bit to see the actual `RequiredError` property because it is hidden by the fixed header of the MudBlazor website. As far as I remember, this can be fixed easily with some frontend magic.

jperson2000 commented:
> Also, there are still bugs reported in the [#8903 (comment)](https://github.com/MudBlazor/MudBlazor/pull/8903#issuecomment-2099194694), and the bug:

I'm not sure if you're reading the whole thread.  As I understand it, there's a release tomorrow, in which case it may be preferable to get to a checkpoint, then continue work on another PR.

iwis commented:
I just wanted to be sure, you are aware of them.
